File: packages/lightning/src/commands/bridge/create.ts
import type { command_opts } from '../../structures/commands.ts';
import { log_error } from '../../structures/errors.ts';
import { bridge_add_common } from './_internal.ts';

export async function create(
	opts: command_opts,
): Promise<string> {
	const result = await bridge_add_common(opts);

	if (typeof result === 'string') return result;

	const bridge_data = {
		name: opts.args.name,
		channels: [result],
		settings: {
			allow_editing: true,
			allow_everyone: false,
			use_rawname: false,
		},
	};

	try {
		const { id } = await opts.bridge_data.create_bridge(bridge_data);
		return `Bridge created successfully!\nYou can now join it using \`${opts.prefix}bridge join ${id}\`.\nKeep this ID safe, don't share it with anyone, and delete this message.`;
	} catch (e) {
		log_error(e, {
			message: 'Failed to insert bridge into database',
			extra: bridge_data,
		});
	}
}

File: internal/commands/create.go
package commands

import (
	"log/slog"

	"github.com/oklog/ulid/v2"
	"github.com/williamhorning/lightning/internal/data"
	"github.com/williamhorning/lightning/pkg/lightning"
)

func bridgeCreate(database data.Database) *lightning.Command {
	return &lightning.Command{
		Name:        "create",
		Description: "create a new bridge in this channel",
		Executor: func(opts *lightning.CommandOptions) {
			channel, err := prepareChannelForBridge(database, opts)
			if err != nil {
				sendErr(err, "failed to setup channel", opts)

				return
			}

			bridge := data.Bridge{
				ID:       ulid.Make().String(),
				Channels: []data.BridgeChannel{*channel},
				Settings: data.BridgeSettings{},
			}

			if err = database.CreateBridge(bridge); err != nil {
				sendErr(err, "failed to save to database", opts)

				return
			}

			if err = opts.Reply(getMessage("created bridge!",
				"you can now join the bridge you made in other channels by using ||`"+opts.Prefix+"bridge join "+
					bridge.ID+"`||. Keep that command secret!"), true); err != nil {
				sendErr(err, "failed to respond to create command", opts)

				bridge.Channels = []data.BridgeChannel{}

				if err = database.CreateBridge(bridge); err != nil {
					slog.Error("failed to remove bridge after failed response", "err", err) // at least log the error
				}
			}
		},
	}
}
